---
title: "Differential Expression Analysis"
author: "Adam Festa"
date: "5/8/2018"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Script objective

The purpose of this script is to conduct differntial expression on the full set of families.  

```{r cars}
suppressPackageStartupMessages(library(DESeq2))
```

## Load data

Load the biological replicate data

```{r pressure, echo=FALSE}
load("~/Projects/RNA_SEQ_Data/shared/Step2_Load_Counts.RData")
```


## Remove data with less than 3 biological replicates

```{r filt.dat}
all.fams <- as.character(load.counts$phenos$fam_id)

remove.fams <- names(which(table(all.fams) < 3))
remove.rows <- which(all.fams %in% remove.fams)

load.counts$txi_object$abundance <- load.counts$txi_object$abundance[,-remove.rows]
load.counts$txi_object$counts <- load.counts$txi_object$counts[,-remove.rows]
load.counts$txi_object$length <- load.counts$txi_object$length[,-remove.rows]
load.counts$txi_object$variance <- load.counts$txi_object$variance[,-remove.rows]

load.counts$phenos <- droplevels.data.frame(load.counts$phenos[-remove.rows,])
load.counts$phenos$fam_id
```

## Add location info

```{r}
 load("~/Projects/RNA_SEQ_Data/disk6directory/get.location.info.RData")
all.fams <- as.character(load.counts$phenos$fam_id)
table(fam.origins$state[sapply(all.fams,function(x) which(fam.origins$family == x))])
state.origins <- fam.origins$state[sapply(all.fams,function(x) which(fam.origins$family == x))]
state.origins[which(state.origins %in% c("AL","TX"))] <- "WG"
table(state.origins)
load.counts$phenos$state <- as.factor(state.origins)
```

## Create DE object

```{r}
ddsTxi <- DESeqDataSetFromTximport(load.counts$txi_object,
                                   colData = load.counts$phenos,
                                   design = ~ batch + index_seq + fam_id)
```


## Run DESeq2

```{r}
library(BiocParallel)
register(MulticoreParam(30))
ddsTxi <- estimateSizeFactors(ddsTxi)
idx <- rowSums( counts(ddsTxi, normalized=TRUE) >= 1 ) >= 206
table(idx)
ddsTxi <- ddsTxi[idx,]

ddsTxi <- DESeq(ddsTxi,parallel = T)
```

Logs:
```
estimating size factors
using 'avgTxLength' from assays(dds), correcting for library size
estimating dispersions
gene-wise dispersion estimates: 30 workers
mean-dispersion relationship
final dispersion estimates, fitting model and testing: 30 workers
```


## Generate all pairwise comparison 

```{r}
all.fams <- as.character(unique(ddsTxi$fam_id))
all.de.txpts <- vector("list")
for(each.fam in 1:length(all.fams)) {
  # Specify the family that everything will be compared to 
  first.comp <- all.fams[each.fam]
  # Specify the list of other families that it will be compared against
  if(each.fam == 1){
  other.comp.fams <- all.fams[-1]
  } else {
   other.comp.fams <- all.fams[-c(1:each.fam)] 
  }
  # Identify DE transcripts for each family
  temp.out <- lapply(1:length(other.comp.fams),function(each.other.fam){ 
    other.fam <- other.comp.fams[each.other.fam]
    results(ddsTxi, 
            contrast = c("fam_id",first.comp,other.fam), 
            parallel = T)
  })
  all.de.txpts[[each.fam]] <- temp.out
  names(all.de.txpts)[[each.fam]] <- first.comp
  print(each.fam)
}
```

### Save Workspace 
 
```{r save.wkspace}
save(all.de.txpts,file = "all.de.txpts.deseq2.Rdata",compress = T)
save(ddsTxi,file = "ddsTxi.deseq2.RData",compress=T)
```
